BREAKING CHANGE: Add generic structure type, use context/getObject

toddbaert opened the pull request:
As described in https://github.com/open-feature/java-sdk/issues/50, I think that expected a `T` back from object flag resolvers is not sustainable. I believe it pushes too much of a burden on providers to instantiate a new T (whatever it may be) and map the attributes received from the backend into it.

Instead, I've added a more generic `Structure` (language used in the OpenFeature spec) with methods similar to the existing `EvaluationContext`. I've also made the `EvaluationContext` inherit from this to reduce duplication (the `EvaluationContext` is also referred to as a "structure" in the spec. 

After this change, providers will create a new `Structure` and add the corresponding fields from the json/xml/protobuf they are concerned with.

In summary:

- Adds `Structure` type which represents JSON data
- `EvaluationContext` ~inherits~ encapsultes this using lombok's `@Delegate`.
- Consolidated `addIntegerAttribute`, `addStringAttribute`, etc, in `EvaluationContext` to a single overloaded `add` method.
- Removed ability to get multiple attributes of a single type from `EvaluationContext` (I'm not convinced these are that valuable, but I'll take feedback).
- introduces a `Value` wrapper class that houses one of: Structure, List<Value>, Boolean, etc (representations of all the possible JSON node values).

beeme1mr commented:
FYI @rgrassian-split

rgrassian-split commented:
> FYI @rgrassian-split. I would like to get your thoughts on this. How would this change impact the provider you're working on?

@beeme1mr Hey! This is a fantastic change, I was actually running into some issues with the `T` implementation so I'm happy to see this. It shouldn't be too crazy to change our provider to support this.

toddbaert commented:
@rgrassian-split it would be good if you could put some feedback in https://github.com/open-feature/java-sdk/issues/50

rgrassian-split commented:
@toddbaert Just did, that is actually the bigger deal for us that I very much support switching it from the current approach

justinabrahms commented on pom.xml:
```diff
@@ -1,5 +1,4 @@
-<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
```
If you would like to change whitespace in this file (or others), please check in a tool which will break the build if it's not right. I'm not tied to my particular formatting choices, but don't want us to fight back and forth about what the right answer should be.

toddbaert replied:
I will see what I can do about this.

justinabrahms commented on pom.xml:
```diff
@@ -246,7 +245,7 @@
 
       <!-- Begin source & javadocs being generated -->
-      <plugin>
+      <!-- <plugin>
```
Reminder that we'll want to uncomment these here and elsewhere.

toddbaert replied:
ugh. I didn't mean to commit this.

I have an issue with this pom and Java 18. I was able to resolve it in the contribs.

toddbaert replied:
reverted.

justinabrahms commented on src/main/java/dev/openfeature/javasdk/EvaluationContext.java:
```diff
@@ -1,139 +1,31 @@
 package dev.openfeature.javasdk;
 
-import dev.openfeature.javasdk.internal.Pair;
-import lombok.*;
```
Similar to above, either make your editor not do these formatting changes or check in something that consistently applies it.

justinabrahms commented on src/main/java/dev/openfeature/javasdk/EvaluationContext.java:
```diff
@@ -1,139 +1,31 @@
-    @Setter @Getter private String targetingKey;
-    private final Map<String, Pair<FlagValueType, Object>> attributes;
+public class EvaluationContext extends Structure {
+    @Setter
+    @Getter
+    private String targetingKey;
```
I'd actually prefer we not split this across 3 lines. `@Setter` isn't enough information to have it's own line, imo.

justinabrahms commented on src/main/java/dev/openfeature/javasdk/Structure.java:
```diff
@@ -0,0 +1,118 @@
+     * @return date time object.
+     * @throws java.time.format.DateTimeParseException if it's not a datetime
+     */
+    public ZonedDateTime getDatetimeAttribute(String key) {
```
Unrelated to your change, but I'm starting to think we should drop this. It feels like it should migrate to contrib, at best.

toddbaert replied:
I think I agree... but there's something like this in the spec. Do you think we need to adjust the spec?

justinabrahms commented on src/test/java/dev/openfeature/javasdk/EvalContextTest.java:
```diff
@@ -1,131 +1,118 @@
-        EvaluationContext out = ec.addStructureAttribute("str", "test")
-                .addIntegerAttribute("int", 4)
-                .addBooleanAttribute("bool", false)
-                .addStructureAttribute("str", new Node<Integer>());
```
You dropped this one from your test.

toddbaert replied:
Added back, thanks!

justinabrahms reviewed (changes requested):
Lots of formatting churn makes the testing stuff (especially spec definitions) a bit hard to read.

justinabrahms commented on src/main/java/dev/openfeature/javasdk/Structure.java:
```diff
@@ -0,0 +1,118 @@
+        return ZonedDateTime.parse(attr, DateTimeFormatter.ISO_ZONED_DATE_TIME);
+    }
+
+    public Structure getStructureAttribute(String key) {
```
We're also missing something like an array type.

toddbaert replied:
I think you are right.

toddbaert replied:
Added

justinabrahms commented on src/main/java/dev/openfeature/javasdk/Structure.java:
```diff
@@ -0,0 +1,118 @@
+@ToString
+@EqualsAndHashCode
+@SuppressWarnings("PMD.BeanMembersShouldSerialize")
+public class Structure {
```
If our backend returns this:

```json
{
  "foo": null
}
```

What type will that serialize into on a structure?

justinabrahms replied:
Provider won't know what type to use. I think the way we work around this is returning `null` for any unknown values. This means if you come across a `null`, you can just not call `add` at all.

toddbaert replied:
> If we don't call add, then foo won't be in the keySet. So... let's just pick something like boolean and shove all nulls in there as a typecasted thing. Users will get nulls out when they ask for them.

OK that makes sense. That's really something for the provider to do though, right? I guess we could add tests for this, but there isn't much for us to do in the SDK to support this, right?

justinabrahms replied:
That's right.

toddbaert replied:
I'll add a specific test for this.

toddbaert replied:
I've added such a test.

toddbaert commented:
@justinabrahms I think I've cleaned up a lot of the whitespace changes. A few are possibly improvements (missing newlines in some places, etc). I've left those.

I also added array support in `Structure`.

I'm going to test this locally in my provider and see if it makes things nicer. Let me know what you think.

justinabrahms commented:
Playing around with the API. It seems like because we broke out `Structure`, the `return this` is returning `Structure` objects, not `EvaluationContext` objects. I'm not aware of a way to have it return `EvaluationContext` objects except to have EC proxy those methods to an internal representation of a `Structure`. Thoughts?

<img width="693" alt="image" src="https://user-images.githubusercontent.com/3853/186283136-004e6578-5b45-49ff-91c5-54cd9dd014a1.png">

toddbaert commented:
> Playing around with the API. It seems like because we broke out `Structure`, the `return this` is returning `Structure` objects, not `EvaluationContext` objects. I'm not aware of a way to have it return `EvaluationContext` objects except to have EC proxy those methods to an internal representation of a `Structure`. Thoughts?

I noticed that too... it's sort of a shame, but I think we'd have to proxy them as well... maybe there's a clever solution but I can't think of one immediately. I'll look around.

toddbaert commented on .github/workflows/pullrequest.yml:
```diff
@@ -27,20 +27,8 @@ jobs:
           restore-keys: |
             ${{ runner.os }}-maven-
 
-      - name: Configure GPG Key
-        if: github.event.pull_request.head.repo.full_name == github.repository
-        run: |
-          echo -n "$GPG_SIGNING_KEY" | base64 --decode | gpg --import
-        env:
-          GPG_SIGNING_KEY: ${{ secrets.GPG_SIGNING_KEY }}
-
       - name: Build with Maven
-        run: |
-          if [ ${{ github.event.pull_request.head.repo.full_name }} = ${{ github.repository }} ]; then
-            mvn --batch-mode --update-snapshots verify
-          else
-            mvn --batch-mode --update-snapshots verify -Dgpg.skip
-          fi
```
Removed signing in PRs.

toddbaert commented on pom.xml:
```diff
@@ -292,7 +292,8 @@
           <execution>
             <id>sign-artifacts</id>
-            <phase>verify</phase>
+            <!-- we only need to sign if we are deploying (publishing) -->
+            <phase>deploy</phase>
```
Only sign in deploy. Nice because now you can run `mvn verify` locally and in PRs without worrying about GPG setup.

justinabrahms commented on src/main/java/dev/openfeature/javasdk/EvaluationContext.java:
```diff
@@ -1,139 +1,31 @@
+@ToString @EqualsAndHashCode(callSuper = true)
 @SuppressWarnings("PMD.BeanMembersShouldSerialize")
-public class EvaluationContext {
+public class EvaluationContext extends Structure {
```
I'm thinking about whether we should be extending structure here, rather than having an instance of Structure as an internal property. I think I have a preference for the latter?

One thing that makes me nervous about this EC/Structure merging is that Array type you added to the flag value types. We don't actually support returning arrays, and so parts of those are leaking between one another.

Thoughts?

justinabrahms replied:
Splitting out the enum will address my concern for FVT. `@Delegate` looks like exactly what I was hoping for. Lombok is the best. Thank you!

toddbaert replied:
Lombok is incredible.

toddbaert replied:
I believe [this commit](https://github.com/open-feature/java-sdk/pull/51/commits/a8623253547411e94fe457adbb1fdd1a973779d7) resolves this concerns.

- encapsulates instead of inherits `Structure`, using lombok's `@Delegate`.
- adds new internal enum for `Structure` field types.

kinyoklion commented on src/main/java/dev/openfeature/javasdk/Structure.java:
```diff
@@ -0,0 +1,137 @@
+     * 
+     * @return all attributes on the structure
+     */
+    public Map<String, Object> getAllAttributes() {
```
Would it be possible to get something like this that includes the type annotations? So that if one is converting this structure into something using another structure, or a builder, that they can iterate it and dispatch the methods to the correct types?

justinabrahms replied:
@kinyoklion I think you should be able to do that with `instanceof`?

toddbaert replied:
> @kinyoklion I think you should be able to do that with instanceof?

Indeed, this is what I'm doing in my provider. Is there some reason it's better to leverage the underlying annotation in the `Pair`?

kinyoklion replied:
Yeah, it can be done that way. Also the performance of `instanceof` seems to be good.

The information was already there, so having to immediately re-construct it, seemed a bit counter intuitive.

justinabrahms replied:
Totally. I just don't want it to be part of a public API

toddbaert commented:
@justinabrahms I'd like to play with this a bit more locally before we merge... I'll let you know if I think any changes are needed. I think this should be sufficient though.

toddbaert commented:
@justinabrahms  I made some more changes. Our abstraction was leaky because any node in a JSON structure can be a JSON structure, an array or a primitive. That wasn't reflected the implementation.

What I've done in [this commit](https://github.com/open-feature/java-sdk/pull/51/commits/ee58654f6bc02fa5e31b0c734fb85a6068e3cda3), is introduce a `Value` wrapper class that houses one of: `Structure`, `List<Value>`, `Boolean`, etc (representations of all the possible JSON node values). The actual value can be retrieved like `value.asBoolean()` or `value.asStructure()`. I believe any JSON structure can be represented this way.

This is similar to how java-grpc represents their structure (which is very similar to JSON) and something like a simplified version of LDValue.

I've maintained most of the previous work. `EvaluationContext` still encapsulates a `Structure`, etc.

toddbaert commented on src/main/java/dev/openfeature/javasdk/EvaluationContext.java:
```diff
@@ -86,7 +86,7 @@ public EvaluationContext add(String key, Structure value) {
     }
 
-    public <T> EvaluationContext add(String key, List<T> value) {
+    public <T extends Value> EvaluationContext add(String key, List<Value> value) {
```
Lists must be of type `Value`.

justinabrahms commented on src/main/java/dev/openfeature/javasdk/FlagEvaluationOptions.java:
```diff
@@ -1,10 +1,13 @@
+import lombok.Singular;
 
-@Value
+@lombok.Value
```
Not clear why this is being qualified rather than importing it as above? Is it because it's only used once?

justinabrahms replied:
NM. It's now ambiguous. I get it.

kinyoklion commented on src/main/java/dev/openfeature/javasdk/Value.java:
```diff
@@ -0,0 +1,206 @@
+    }
+
+    /** 
+     * Check if this Value represents a ZonedDateTime.
```
Because it shares a storage format with string, would it be worthwhile to have a comment that dates are also strings? If you do care about dates, and you are using this in some kind of dispatch, then you would want to check isZonedDateTime before isString, or check both, or something to disambiguate.

toddbaert replied:
I'm wondering if it isn't better to simply store a datetime... Storing them as strings only adds this complication, as well as performance issues around parsing/error handling. It was stored as a string in the previous implementation and I didn't give it much thought when I translated it.

kinyoklion replied:
I think storing it as a datetime, or removing it, are both good options.

toddbaert replied:
Now storing as ZonedDateTIme.

justinabrahms commented on src/main/java/dev/openfeature/javasdk/Value.java:
```diff
@@ -0,0 +1,206 @@
+@ToString
+@EqualsAndHashCode
+@SuppressWarnings("PMD.BeanMembersShouldSerialize")
+public class Value {
```
```suggestion
public class StructureValue {
```
?

toddbaert replied:
Not sure I agree with this one. I would be OK with `Node`, but I think `Value` is better.

Adding "Structure" seems wrong, since the value can be a primitive.

justinabrahms reviewed (commented):
I think I need to play around with constructing a few of these objects to figure out what the API feels like.

kinyoklion commented on src/main/java/dev/openfeature/javasdk/internal/StructureFieldType.java:
```diff
@@ -0,0 +1,5 @@
+package dev.openfeature.javasdk.internal;
+
+public enum StructureFieldType {
```
This isn't used now that there pairs were removed?

toddbaert replied:
Yep, great catch.

toddbaert replied:
Removed.

toddbaert commented:
> I think I need to play around with constructing a few of these objects to figure out what the API feels like.

OK @justinabrahms, I'll leave it up to you to merge it then (or message with more feedback). @kinyoklion has also approved. I think we'll need to do something similar in dotnet cc @benjiro

s-sen commented on src/main/java/dev/openfeature/javasdk/Structure.java:
```diff
@@ -0,0 +1,87 @@
+     * @return all attributes on the structure
+     */
+    public Map<String, Value> asMap() {
+        return new HashMap<>(this.attributes);
```
Are we OK with a shallow copy of the values? The application developer will use this to obtain all the values hence a  shallow copy is OK in that scenario. But will this structure also be used during building and mutating the EvaluationContext?

toddbaert replied:
I think this is a good question...

I suspect that for mutating, users will be using the `merge` and `add` functions in the EvaluationContext. I don't have any strong opinions about if this should be a shallow copy, deep copy, or just expose the inner map (though I think in most cases, that's a bad choice). I guess whatever we choose it should be well-documented.

Does anyone have a strong argument for this?

@justinabrahms @kinyoklion

justinabrahms replied:
https://github.com/open-feature/java-sdk/issues/53 Moved this discussion here.
